### PR TITLE
RDM-4544 - validate regular expression for document

### PR DESCRIPTION
**Pull request by gbenadikar**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/RDM-4544

### Change description ###
Validate regular expression for document field type to accept a valid pattern

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: domain/src/main/java/uk/gov/hmcts/ccd/definition/store/domain/validation/complexfield/ComplexFieldEntityRegularExpressionValidator.java [hunk 1]
```diff
@@ -0,0 +1,18 @@
+package uk.gov.hmcts.ccd.definition.store.domain.validation.complexfield;
+
+import org.springframework.stereotype.Component;
+import uk.gov.hmcts.ccd.definition.store.domain.validation.ValidationResult;
+import uk.gov.hmcts.ccd.definition.store.domain.validation.regex.AbstractRegularExpressionValidator;
+import uk.gov.hmcts.ccd.definition.store.repository.entity.ComplexFieldEntity;
+
+@Component
+public class ComplexFieldEntityRegularExpressionValidator extends AbstractRegularExpressionValidator implements ComplexFieldValidator {
```
Comment (mario-paniccia) [inline domain/src/main/java/uk/gov/hmcts/ccd/definition/store/domain/validation/complexfield/ComplexFieldEntityRegularExpressionValidator.java]: the code in ComplexFieldEntityRegularExpressionValidator and AbstractRegularExpressionValidator is very specific to Document fields and regex validation on Document fields. So it doesn't look like it's reusable for other complex types and other types of validation on complex types. I would then suggest to remove these 2 classes and merge that code into 1 class `DocumentRegexValidatator` which extends `ComplexFieldValidator`

Comment (gbenadikar) [reply domain/src/main/java/uk/gov/hmcts/ccd/definition/store/domain/validation/complexfield/ComplexFieldEntityRegularExpressionValidator.java]: FieldTypeRegularExpressionValidator and ComplexFieldEntityRegularExpressionValidator both extend AbstractRegularExpressionValidator as both need to do that validation. I kept the name generic and not specific to doc type. I can rename all 3 specific to Document Type for e.g. DocTypeRegularExpressionValidator, ComplexFieldDocTypeRegularExpressionValidator and AbstractDocTypeRegularExpressionValidator

Comment (mario-paniccia) [reply domain/src/main/java/uk/gov/hmcts/ccd/definition/store/domain/validation/complexfield/ComplexFieldEntityRegularExpressionValidator.java]: what I'm saying is that even if you kept the name generic, their code is tied to Documents and Regex validation so it's not generic

Comment (gbenadikar) [reply domain/src/main/java/uk/gov/hmcts/ccd/definition/store/domain/validation/complexfield/ComplexFieldEntityRegularExpressionValidator.java]: Renamed

File: domain/src/main/java/uk/gov/hmcts/ccd/definition/store/domain/validation/regex/DocumentTypeRegularExpressionValidationError.java [hunk 2]
```diff
@@ -0,0 +1,14 @@
+package uk.gov.hmcts.ccd.definition.store.domain.validation.regex;
+
+import uk.gov.hmcts.ccd.definition.store.domain.validation.ValidationError;
+import uk.gov.hmcts.ccd.definition.store.repository.entity.FieldTypeEntity;
+
+public class DocumentTypeRegularExpressionValidationError extends ValidationError {
```
Comment (mario-paniccia) [inline domain/src/main/java/uk/gov/hmcts/ccd/definition/store/domain/validation/regex/DocumentTypeRegularExpressionValidationError.java]: looking at the code of this one, if we change:
`Invalid regular expression for Document field...` to `Invalid regular expression for field...`
then this one is generic and could be used for all regex exceptions on any field type. So we could name it `RegularExpressionValidationError`

Comment (gbenadikar) [reply domain/src/main/java/uk/gov/hmcts/ccd/definition/store/domain/validation/regex/DocumentTypeRegularExpressionValidationError.java]: done

**Pull request merged**